### PR TITLE
CP-696 Waterline changes to use new Windward Gateway

### PR DESCRIPTION
**Pull request by konradkurze**
Waterline changes to use new Windward Gateway
Jeffrey Henley

### CODE REVIEW

File: lib/Connection.js [hunk 1]
```diff
@@ -43,29 +35,59 @@ class Connection {
...
+				if (this.isInvalidSfSession(response.status, body)) {
+					tries = tries || 0;
+					if (tries > 2) {
+						return Promise.reject(new Error(INVALID_RESPONSE));
+					}
+
+					//Renew Access Token and try again
+					return this.getAccessToken()
+						.then(() => {
+							return this.fetch(`${this.instanceUrl}${path}`, options, ++tries);
```
Comment (cleathers) [inline lib/Connection.js]: Why `++tries` instead of `tries++`?

File: lib/Connection.js [hunk 2]
```diff
@@ -43,29 +35,59 @@ class Connection {
...
-	 * validates access token and then returns connection details
-	 * @returns {Promise.<T>}
-	 */
-	getConnection() {
-		return this.validateAccessToken()
-			.catch(() => {
+	isInvalidSfSession(statusCode, body) {
+		var error = this.parseError(body);
+		error = typeof error === 'object' ? error : {};
+		return (statusCode === 401 && error.errorCode === 'INVALID_SESSION_ID');
```
Comment (cleathers) [inline lib/Connection.js]: no need for the parens here

File: lib/Connection.js [hunk 3]
```diff
@@ -92,6 +113,7 @@ class Connection {
 			.then((response) => {
 				if (!response.ok) {
 					console.warn(`${LOG_ID}getAccessToken bad response from API Gateway`);
+					console.log('response.status: ', response.status);
```
Comment (cleathers) [inline lib/Connection.js]: Do we still need this new logging or is the `warn` from before enough?

**Comment by cleathers** [comment]
Looks good as far as I can tell, just a couple small questions

File: test/unit/Connection.test.js [hunk 4]
```diff
@@ -331,13 +257,13 @@ test('Connection::fetch', (t) => {
 	var fetchSpy = sinon.spy();
 	var fakeFetch = (url, options) => {
 		fetchSpy.apply(null, [url, options]);
-		return Promise.resolve();
+		return Promise.resolve({fakeResponse: true, status: 200});
```
Comment (camsjams) [inline test/unit/Connection.test.js]: where is `fakeResponse` used?

File: test/unit/Connection.test.js [hunk 5]
```diff
@@ -53,6 +55,24 @@ var Connection = proxyquireStrict('../../lib/Connection.js', {
 	aws4: aws4Mock
 });
 
+var altAwsSpy = sinon.spy();
+var fileSysMock = sinon.spy();
+var altAwsMock = {
+	config: {},
+	EC2MetadataCredentials: function mockEC2MetadataCredentials() {
+		altAwsSpy();
```
Comment (camsjams) [inline test/unit/Connection.test.js]: should be a sinon `stub`

File: test/unit/Connection.test.js [hunk 6]
```diff
@@ -53,6 +55,24 @@ var Connection = proxyquireStrict('../../lib/Connection.js', {
...
+		return awsCredentials;
+	},
+	FileSystemCredentials: function mockFileSystemCredentials(path) {
+		fileSysMock(path);
```
Comment (camsjams) [inline test/unit/Connection.test.js]: should be a sinon `stub`

File: test/unit/Connection.test.js [hunk 7]
```diff
@@ -156,8 +123,7 @@ test('Connection::getAccessToken', (t) => {
 		[EXPECTED_ADDED_HEADER]: EXPECTED_ADDED_HEADER_VALUE
 	};
 	const EXPECTED_BODY = JSON.stringify({
-		service: 'salesforce',
-		env: FAKE_CREDENTIALS.env
+		service: 'salesforce'
```
Comment (camsjams) [inline test/unit/Connection.test.js]: 👍

File: lib/Connection.js (hunk 1 above)
Comment (konradkurze) [inline lib/Connection.js]: tries++ will pass back tries as is then increment it after, ++tries increments then passes (so tries++ will pass back the same number each time as it'll pass back the original then increment the local var.

File: lib/Connection.js (hunk 3 above)
Comment (konradkurze) [inline lib/Connection.js]: Oops. was a testing comment

File: test/unit/Connection.test.js (hunk 4 above)
Comment (konradkurze) [inline test/unit/Connection.test.js]: No where, it's just marking it as a fake response. Can take it out if you think it just adds clutter.

Comment (camsjams) [inline test/unit/Connection.test.js]: minor clutter, up to you

**Pull request merged**